File: spotify_helper.py
import concurrent.futures
import os
import re

import spotipy
from spotipy.oauth2 import SpotifyOAuth, SpotifyClientCredentials

SPOTIFY_USER_ID = os.environ["SPOTIFY_USER_ID"]

SPOTIFY_SCOPE = "user-library-read playlist-modify-public"

# Spotify API limit for track IDs per request
MAX_IDS_PER_REQUEST = 50

def chunk_list(lst, chunk_size):
    """Yield successive n-sized chunks from lst."""
    for i in range(0, len(lst), chunk_size):
        yield lst[i:i + chunk_size]

class SpotifyAPIHelper:
    def __init__(self):
        # self._auth_manager = SpotifyClientCredentials()
        # self._sp = spotipy.Spotify(auth_manager=self._auth_manager)
        
        self._sp = spotipy.Spotify(auth_manager=SpotifyOAuth(scope=SPOTIFY_SCOPE))

    def _playlist_exists(self, name):
        playlists = self._sp.user_playlists(SPOTIFY_USER_ID)
        for p in playlists["items"]:
            if p["name"] == name:
                return p["id"]

        return None
    
    def search_track(self, name):
        q = f"track:{name}"
        results = self._sp.search(q)
        
        if "tracks" in results and len(results["tracks"]["items"]) > 0:
            return results["tracks"]["items"][0]["external_urls"]["spotify"]
        else:
            return None
     
    def _create_playlist(self, name):
        playlist = self._sp.user_playlist_create(SPOTIFY_USER_ID, name)
        return playlist["id"]
    
    def update_playlist(self, name, track_uris):
        playlist_id = self._playlist_exists(name)
        existing_tracks = set()
        if not playlist_id:
            print(f"Creating playlist {name}")
            playlist_id = self._create_playlist(name)
        else:
            playlist_items = self._sp.playlist_tracks(playlist_id)
            for item in playlist_items["items"]:
                existing_tracks.add(item["track"]["external_urls"]["spotify"])
    
        # print(f"Tracks to be added: {track_uris}\n\n")
        # print(f"Existing tracks: {existing_tracks}\n\n")
        new_tracks = set(track_uris) - existing_tracks
            
        if len(new_tracks) > 0:
            print(f"Adding tracks {new_tracks} to playlist {playlist_id}")
            self._sp.playlist_add_items(playlist_id, new_tracks)

    def _extract_spotify_url_bits(self, links):
        # Regular expression to explicitly match URLs starting with 'https://open.spotify.com/'
        pattern = r'https://open\.spotify\.com/([^/]+)/([^?]+)'

        ret = []        
        for url in links:
            if url.find("spotify") < 0:
                continue

            # Search for matches
            match = re.search(pattern, url)
            
            if match:
                word = match.group(1)  # The word (e.g., "playlist")
                id_value = match.group(2)  # The ID (e.g., "37i9dQZF1DWYCFWZy4Gz9M")
                ret.append({"url": url, "type": word, "id": id_value})
            else:
                print(f"Could not parse Spotify URL: {url}")
                
        return ret
    
    def _get_tracks(self, track_ids):
        ret = {}
        
        if len(track_ids) > 0:            
            for chunk in chunk_list(list(track_ids.keys()), MAX_IDS_PER_REQUEST):    
                tracks = self._sp.tracks(chunk)
                for track in tracks["tracks"]:
                    track_name = track["name"]
                    artists = ", ".join([a["name"] for a in track["artists"]])
                    
                    link_url = track_ids[track["id"]]
                    ret[link_url] = {
                        "name": track_name, 
                        "artists": artists,
                        "type": "track"
                    }
                    
                    images = track["album"]["images"]
                    if len(images) > 0:
                        ret[link_url]["image_url"] = images[0]["url"]

        return ret
    
    def _get_playlists(self, playlist_ids):
        ret = {}
        
        for playlist_id in playlist_ids:
            playlist = self._sp.playlist(playlist_id)
            
            link_url = playlist_ids[playlist_id]
            ret[link_url] = {
                "name": playlist["name"], 
                "type": "playlist"
            }
            
            if len(playlist["images"]) > 0:
                ret[link_url]["image_url"] = playlist["images"][0]["url"]
        
        return ret
    
    def _get_albums(self, album_ids):
        ret = {}

        if len(album_ids) > 0:
            albums = self._sp.albums(list(album_ids.keys()))
            for album in albums["albums"]:
                album_name = album["name"]
                artists = ", ".join([a["name"] for a in album["artists"]])
                
                link_url = album_ids[album["id"]]
                ret[link_url] = {
                    "name": album_name, 
                    "artists": artists,
                    "type": "album"
                }
                
                if len(album["images"]) > 0:
                    ret[link_url]["image_url"] = album["images"][0]["url"]
 
        return ret

    def _get_artists(self, artist_ids):
        ret = {}

        if len(artist_ids) > 0:
            artists = self._sp.artists(list(artist_ids.keys()))
            for artist in artists["artists"]:
                artist_name = artist["name"]
                
                link_url = artist_ids[artist["id"]]
                ret[link_url] = {
                    "name": artist_name, 
                    "type": "artist"
                }
                
                if len(artist["images"]) > 0:
                    ret[link_url]["image_url"] = artist["images"][0]["url"]
 
        return ret

    def _parallel_get_data(self, track_ids, album_ids, playlist_ids, artist_ids):
        # Create a ThreadPoolExecutor to manage threads
        with concurrent.futures.ThreadPoolExecutor() as executor:
            # Submit all four tasks to the executor
            future_tracks = executor.submit(self._get_tracks, track_ids)
            future_albums = executor.submit(self._get_albums, album_ids)
            future_playlists = executor.submit(self._get_playlists, playlist_ids)
            future_artists = executor.submit(self._get_artists, artist_ids)
            
            # Wait for all the futures to complete and collect the results
            ret = {}
            
            # Combine the results into the ret dictionary
            ret.update(future_tracks.result())
            ret.update(future_albums.result())
            ret.update(future_playlists.result())
            ret.update(future_artists.result())
            
        # Return the combined result
        return ret
      
    def analyse_spotify(self, links):
        # For each Spotify link, get ID and link type.
        spotify_bits = self._extract_spotify_url_bits(links)

        track_ids = {}
        playlist_ids = {}
        album_ids = {}
        artist_ids = {}        
        for row in spotify_bits:
            type = row["type"]
            id = row["id"]
            url = row["url"]
            
            if type == "track":
                track_ids[id] = url
            elif type == "album":
                album_ids[id] = url
            elif type == "playlist":
                playlist_ids[id] = url
            elif type == "artist":
                artist_ids[id] = url
            else:
                print(f"Unknown type: {type}")
        
        return self._parallel_get_data(
            track_ids, album_ids, playlist_ids, artist_ids)
                
            

File: client_only/spotify_helper.js
const SpotifyWebApi = require('spotify-web-api-node');

const SPOTIFY_USER_ID = process.env.SPOTIFY_USER_ID;
const SPOTIFY_SCOPE = "user-library-read playlist-modify-public";

// Spotify API limit for track IDs per request
const MAX_IDS_PER_REQUEST = 50;

function chunkList(lst, chunkSize) {
  const chunks = [];
  for (let i = 0; i < lst.length; i += chunkSize) {
    chunks.push(lst.slice(i, i + chunkSize));
  }
  return chunks;
}

class SpotifyAPIHelper {
  constructor() {
    this._sp = new SpotifyWebApi({
      clientId: process.env.SPOTIFY_CLIENT_ID,
      clientSecret: process.env.SPOTIFY_CLIENT_SECRET,
      redirectUri: process.env.SPOTIFY_REDIRECT_URI,
      refreshToken: process.env.SPOTIFY_REFRESH_TOKEN
    });

    this._sp.setAccessToken(process.env.SPOTIFY_ACCESS_TOKEN);
    
    // Refresh token on initialization
    this._refreshAccessToken();
  }
  
  async _refreshAccessToken() {
    try {
      const data = await this._sp.refreshAccessToken();
      this._sp.setAccessToken(data.body.access_token);
    } catch (error) {
      console.error('Could not refresh access token', error);
    }
  }

  async _playlistExists(name) {
    try {
      const data = await this._sp.getUserPlaylists(SPOTIFY_USER_ID);
      for (const p of data.body.items) {
        if (p.name === name) {
          return p.id;
        }
      }
      return null;
    } catch (error) {
      console.error('Error checking if playlist exists:', error);
      return null;
    }
  }
  
  async searchTrack(name) {
    try {
      const q = `track:${name}`;
      const results = await this._sp.searchTracks(q);
      
      if (results.body.tracks && results.body.tracks.items.length > 0) {
        return results.body.tracks.items[0].external_urls.spotify;
      } else {
        return null;
      }
    } catch (error) {
      console.error('Error searching track:', error);
      return null;
    }
  }
     
  async _createPlaylist(name) {
    try {
      const playlist = await this._sp.createPlaylist(SPOTIFY_USER_ID, {
        name: name,
        public: true
      });
      return playlist.body.id;
    } catch (error) {
      console.error('Error creating playlist:', error);
      return null;
    }
  }
  
  async updatePlaylist(name, trackUris) {
    try {
      let playlistId = await this._playlistExists(name);
      const existingTracks = new Set();
      
      if (!playlistId) {
        console.log(`Creating playlist ${name}`);
        playlistId = await this._createPlaylist(name);
      } else {
        const playlistItems = await this._sp.getPlaylistTracks(playlistId);
        for (const item of playlistItems.body.items) {
          existingTracks.add(item.track.external_urls.spotify);
        }
      }
    
      const newTracks = trackUris.filter(uri => !existingTracks.has(uri));
            
      if (newTracks.length > 0) {
        console.log(`Adding tracks ${newTracks} to playlist ${playlistId}`);
        await this._sp.addTracksToPlaylist(playlistId, newTracks);
      }
    } catch (error) {
      console.error('Error updating playlist:', error);
    }
  }

  _extractSpotifyUrlBits(links) {
    // Regular expression to explicitly match URLs starting with 'https://open.spotify.com/'
    const pattern = /https:\/\/open\.spotify\.com\/([^/]+)\/([^?]+)/;

    const ret = [];        
    for (const url of links) {
      if (url.indexOf("spotify") < 0) {
        continue;
      }

      // Search for matches
      const match = pattern.exec(url);
      
      if (match) {
        const word = match[1];  // The word (e.g., "playlist")
        const idValue = match[2];  // The ID (e.g., "37i9dQZF1DWYCFWZy4Gz9M")
        ret.push({ url, type: word, id: idValue });
      } else {
        console.log(`Could not parse Spotify URL: ${url}`);
      }
    }
    
    return ret;
  }
  
  async _getTracks(trackIds) {
    const ret = {};
    
    if (Object.keys(trackIds).length > 0) {
      try {
        for (const chunk of chunkList(Object.keys(trackIds), MAX_IDS_PER_REQUEST)) {    
          const response = await this._sp.getTracks(chunk);
          for (const track of response.body.tracks) {
            const trackName = track.name;
            const artists = track.artists.map(a => a.name).join(", ");
            
            const linkUrl = trackIds[track.id];
            ret[linkUrl] = {
              name: trackName, 
              artists: artists,
              type: "track"
            };
            
            const images = track.album.images;
            if (images.length > 0) {
              ret[linkUrl].image_url = images[0].url;
            }
          }
        }
      } catch (error) {
        console.error('Error getting tracks:', error);
      }
    }

    return ret;
  }
  
  async _getPlaylists(playlistIds) {
    const ret = {};
    
    for (const playlistId of Object.keys(playlistIds)) {
      try {
        const playlist = await this._sp.getPlaylist(playlistId);
        
        const linkUrl = playlistIds[playlistId];
        ret[linkUrl] = {
          name: playlist.body.name, 
          type: "playlist"
        };
        
        if (playlist.body.images.length > 0) {
          ret[linkUrl].image_url = playlist.body.images[0].url;
        }
      } catch (error) {
        console.log(`Could not find playlist ${playlistId}: ${error.message}`);
      }
    }
    
    return ret;
  }
  
  async _getAlbums(albumIds) {
    const ret = {};

    if (Object.keys(albumIds).length > 0) {
      try {
        const response = await this._sp.getAlbums(Object.keys(albumIds));
        for (const album of response.body.albums) {
          const albumName = album.name;
          const artists = album.artists.map(a => a.name).join(", ");
          
          const linkUrl = albumIds[album.id];
          ret[linkUrl] = {
            name: albumName, 
            artists: artists,
            type: "album"
          };
          
          if (album.images.length > 0) {
            ret[linkUrl].image_url = album.images[0].url;
          }
        }
      } catch (error) {
        console.error('Error getting albums:', error);
      }
    }

    return ret;
  }

  async _getArtists(artistIds) {
    const ret = {};

    if (Object.keys(artistIds).length > 0) {
      try {
        const response = await this._sp.getArtists(Object.keys(artistIds));
        for (const artist of response.body.artists) {
          const artistName = artist.name;
          
          const linkUrl = artistIds[artist.id];
          ret[linkUrl] = {
            name: artistName, 
            type: "artist"
          };
          
          if (artist.images.length > 0) {
            ret[linkUrl].image_url = artist.images[0].url;
          }
        }
      } catch (error) {
        console.error('Error getting artists:', error);
      }
    }

    return ret;
  }

  async _parallelGetData(trackIds, albumIds, playlistIds, artistIds) {
    // Run all API calls in parallel using Promise.all
    const [tracks, albums, playlists, artists] = await Promise.all([
      this._getTracks(trackIds),
      this._getAlbums(albumIds),
      this._getPlaylists(playlistIds),
      this._getArtists(artistIds)
    ]);
    
    // Combine the results
    return {
      ...tracks,
      ...albums,
      ...playlists,
      ...artists
    };
  }
      
  async analyseSpotify(links) {
    // For each Spotify link, get ID and link type.
    const spotifyBits = this._extractSpotifyUrlBits(links);

    const trackIds = {};
    const playlistIds = {};
    const albumIds = {};
    const artistIds = {};        
    for (const row of spotifyBits) {
      const type = row.type;
      const id = row.id;
      const url = row.url;
      
      if (type === "track") {
        trackIds[id] = url;
      } else if (type === "album") {
        albumIds[id] = url;
      } else if (type === "playlist") {
        playlistIds[id] = url;
      } else if (type === "artist") {
        artistIds[id] = url;
      } else {
        console.log(`Unknown type: ${type}`);
      }
    }
    
    return await this._parallelGetData(
      trackIds, albumIds, playlistIds, artistIds);
  }
}

module.exports = {
  SpotifyAPIHelper,
  chunkList
};